syntax = "proto3";

service RaftNodeService {
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  rpc GetLeader(GetLeaderRequest) returns (GetLeaderResponse);
  rpc InstallSnapshot(InstallSnapshotRequest) returns (InstallSnapshotResponse);
}

message AppendEntriesRequest {
  int32 term = 1;
  int32 leader_id = 2;
  int32 prev_log_index = 3;
  int32 prev_log_term = 4;
  repeated LogEntry entries = 5;
  int32 leader_commit = 6;
}

message LogEntry {
  int32 term = 1;
  bytes data = 2;
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
  int32 conflict_index = 3;
  int32 conflict_term = 4;
}

message RequestVoteRequest {
  int32 term = 1;
  int32 candidate_id = 2;
  int32 last_log_index = 3;
  int32 last_log_term = 4;
}

message RequestVoteResponse {
  int32 term = 1;
  bool vote_granted = 2;
}

message GetLeaderRequest {}

message GetLeaderResponse {
  string leader_address = 1;
}

message InstallSnapshotRequest {
  int32 term = 1;
  int32 leader_id = 2;
  int32 last_included_index = 3;
  int32 last_included_term = 4;
  bytes chunk = 5;
  bool done = 6;
}

message InstallSnapshotResponse {
  int32 term = 1;
}